core/src/main/java/org/apache/iceberg/io/WriterFactory.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.deletes.EqualityDeleteWriter;
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+
+/**
+ * A factory for creating data and delete writers.
+ */
+public interface WriterFactory<T> {
+
+  /**
+   * Creates a new {@link DataWriter}.
+   *
+   * @param file the output file
+   * @param spec the partition spec written data belongs to
```

Comment:
Is there possible that we will write data/delete files to an existing historical partition spec ?   I mean almost all the cases, we will produce data/delete files in the latests partition spec,  so when opening new writers among different parallelize tasks,  there seems no need to pass an extra partition spec with the same value.

Reply:
There is actually a use case for that in Spark and I think it applies to other engines too. Imagine we have a table with multiple specs. In Spark, we plan to project `_spec` and `_partition` metadata columns. Whenever we a delete file, we have to use the partition spec the referenced data rows belong to.

Reply:
I think I can understand the point that we want to write the delete files and data files in the same partition for a given partition spec.  But I still don't get the point what's the use case that spark will write data files or delete files into an historical partition spec. Are there any other issues or PRs that I missed ?

Reply:
Consider the following sequence of actions:
1. create table (ts, category, data) partitioned by days(ts)
2. day 1 adds some data files
3. alter table add partition field category
4. day 2 some data files
5. delete some data where day=1 and data=xxx, in this case the delete writer should use the old partition spec. If writing using the latest spec, it would produce 1 delete file per category.

Reply:
OK,  @jackye1995 's comment  answered my question perfectly.  Make sense !

Reply:
@jackye1995's example is correct. The reason why we need to project `_spec` and `_partition` for all rows is that we may need to write using multiple specs within a txn. In the considered example, we need to use the old spec for deleting records in DAY 1 and we may need to use the current spec for deleting records in DAY 2 (all records for DAY 2 are actually written using the new spec).

Reply:
To sum up, the spec for deletes should match the data spec of rows we reference.

Reply:
Another thing to keep in mind is that when encoding deletes, we need to encode them for all of the partition specs that are in use in the table. In @jackye1995's example, I think the predicate needs to be added to `spec=1/day=1` and `spec=2/day=1/category=*` because deletes aren't scoped to just the current spec and there could be data in the new spec.

Reply:
@rdblue, that applies to equality deletes and upsert use cases, right? In case of MERGE INTO, we know the source row spec id and its partition value?

Reply:
One more thing: the deletes can also be added to the unpartitioned spec for global deletes. That avoids the need to find all of the categories that exist in this example.